### ISSUE TITLE
How we can support other parsers better (babel-eslint, typescript)

### ISSUE THREAD

**Issue by hzoo** [eslint/eslint#6364]
> There's a good amount of previous discussion in https://github.com/eslint/eslint/issues/4640 and part of the [wish list issue](https://github.com/eslint/eslint/issues/5040#issuecomment-177569407), but wanted to bring up an issue for this in general.

Just wanted to brainstorm what we can do to make it easier to support experimental syntax or syntax/nodes that core rules might not know about.
- [x] There's currently monkeypatching being done on `VisitorKeys` for `estraverse` to account for unknown node types as well as for `escope` to add variable references and visit them so not sure what we can do there except figure out how if we have add more formal hooks?
- [ ] And there's the issue with `eslint-plugin-babel` which everyone eventually stumbles upon (eslint issue -> babel-eslint -> eslint-plugin-babel). Because of the use of new syntax, users will still want to use old rules. However those rules may error out when it sees new syntax. This leads to https://github.com/babel/babel-eslint/issues/131 where either babel-eslint needs to convert the experimental syntax to something eslint rules understand (and thus the AST is different and unexpected for the rules that try to operate on the new syntax) or we copy paste a specific version of the rule and modify a line or two.

Example: I think the main difference for `arrow-parens` is the line https://github.com/babel/eslint-plugin-babel/blob/master/rules/arrow-parens.js#L24 to account for `node.async`: `if (node.async) token = context.getTokenAfter(token);`. However the rule can get updated upstream and will be out of sync (like in https://github.com/babel/eslint-plugin-babel/issues/63).

How can we extend built-in rules, extend on top?
- [ ] And other parsers need to figure out a good way to run all eslint tests or eslint add integration tests (since we got https://github.com/babel/babel-eslint/issues/267 before) - there were some updates to the tester so just need to further investigate the issues [there](https://github.com/babel/babel-eslint/issues/62).

**Comment by platinumazure** [comment]
Regarding monkey-patching of estraverse/escope, I think #5543 (plus upstream PRs in estraverse and escope) has fixed that? ESLint at least no longer needs to patch estraverse/escope directly, so I guess the question becomes whether or not it is possible for babel-eslint to provide its own overrides?

**Comment by ilyavolodin** [comment]
@hzoo I though #5543 took care of the need for monkey-patching. Is that not correct? Regarding extending rules, it's been brought up multiple times before (mostly by me), but I don't think anyone was able to come up with a good pattern we can use for cases like that. It would be great if we could do it, I just don't know how.

**Comment by mysticatea** [comment]
A small idea, it's to add an event for parsers.
My image is:

```js
function addressARule(context) {
    var sourceCode = context.getSourceCode();
    var token = sourceCode.getFirstToken(context.reportedItem.loc);
    if (node.type === "Keyword" && node.value === "async") {
        // It can cancel this report.
        context.cancel();

        // Or it can modify this report.
        context.reportedItem.loc.column += 5;
    }
}

module.exports = {
    parse: function() {},

    // This is called in RuleContext#report(), then this modifies the report.
    onReported: function(context) {
        switch (context.ruleId) {
            case "a-rule": addressARule(context); break;
            case "another-rule": addressAnotherRule(context); break;
        }
    }
}
````

I think such event have smaller risk than monkey patching.

**Comment by danez** [comment]
Seems that the changes to in #5543 were only made to support unknown node types. So in case of nodes that are known by [estraverse](https://github.com/estools/estraverse/blob/master/estraverse.js#L218) you need to overwrite the internal VisitorKeys, as the `fallback` will only be used when there is [no entry in the VisitorKeys](https://github.com/estools/estraverse/blob/master/estraverse.js#L517). One can overwrite the keys in the estraverse API but only if you are the one calling `traverse()` which eslint is doing somewhere internally. And babel-eslint needs to change VisitorKeys for example for `decorators`.

In the case or babel-eslint I don't see any other way currently besides monkeypatching. The only other solution would be if eslint would allow the parser to supply custom VisitorKeys. 
But maybe I'm also missing something.

**Comment by kaicataldo** [comment]
I think this is still a worthwhile discussion, particularly in light of the work being done on the TypeScript parser.

Here's one thing currently being worked on to ease the burden of reimplementing the comment attachment algorithm in Espree in all parsers: https://github.com/eslint/eslint/issues/6724

**Comment by soda0289** [comment]
Now that we have escope forked as eslint-scope I think we can offer a lot more improvements. @danez brings up some good points about not being able to modify the traversal or scoping of know types. I think it would be nice if the parser could expose their own list of VisitorKeys and Syntax strings that could be passed into the esint-scope ScopeManager and estraverse Traverser. We already have the ability for parsers to expose extra services https://github.com/eslint/eslint/issues/6974 and we could have something similar for Visitor details.

@nzakas has suggested traversing into type annotations and decorators: https://github.com/eslint/eslint/issues/7129 . But this still does not solve all the problems parsers will have. For example in typescript there are Abstract Classes that need to be scoped the same as Classes and Namespaces that need their own special handling. It would be nice if the typescript-eslint-parser could modify the escope Referencer class by extending it. This way the parser could call into the ScopeManger to have it scoped correctly. It could also overload method calls for VisitorKeys such as classes or functions. This could provide the ability to scope decorators, type annotations, or any other construct better and without the need to update eslint. Overall this would give us the same ability as monkey patching does in babel-eslint with a more defined API of the new eslint-scope.

In summary if we allow parsers, such as typescript-eslint-parser and babel-eslint, to provide VisitorKeys and a Refenecer/Visitor instance it would allow for more comprehensive scoping and limit the need to update eslint or eslint-scope with knowledge of these unkown elements.

@JamesHenry @hzoo Would this help with parsing issues or increase the complexity of a parser too much?

**Comment by nzakas** [comment]
> In summary if we allow parsers, such as typescript-eslint-parser and babel-eslint, to provide VisitorKeys and a Refenecer/Visitor instance it would allow for more comprehensive scoping and limit the need to update eslint or eslint-scope with knowledge of these unkown elements.

I agree, that's part of why we are forking escope. However, I don't know that we want to stick with escope's current structure and so I think it's too soon to talk about allowing parsers to expose Referencer/Visitor/other objects. I'd like to think that we can design a cleaner and more intuitive interface to allow parsers to expose more information about the language they are parsing. 

Long term, we definitely want all the hacks in babel-ESLint to go away, I just think we need to take some time to really think through how we want to do that.

**Comment by hzoo** [comment]
> Long term, we definitely want all the hacks in babel-ESLint to go away, I just think we need to take some time to really think through how we want to do that.

Yes 😍  🙏 ! https://github.com/babel/babylon/pull/277 will help with getting rid those specific AST node changes. Right the next thing is the scoping/traversal code for new node types

**Comment by not-an-aardvark** [comment]
I just published [eslint-rule-composer](https://github.com/not-an-aardvark/eslint-rule-composer), which should hopefully make it easier to create variants of rules in `eslint-plugin-babel` without copy-pasting and changing a line.

**Comment by not-an-aardvark** [comment]
Closing this issue since there hasn't been any discussion on it for awhile, but we're certainly still open to individual suggestions for better parser support.
